Add selected dataset ID to URL for sharing

dbczumar opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/dbczumar/mlflow/pull/19142?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19142/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19142/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19142/merge
```

</p>
</details>

<!-- Remove unused checkboxes except for the patch release section at the bottom -->

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Add selected dataset ID to URL for sharing

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

danielseong1 commented on mlflow/server/js/src/experiment-tracking/pages/experiment-evaluation-datasets/components/ExperimentEvaluationDatasetsListTable.tsx:
```diff
@@ -215,17 +213,29 @@ export const ExperimentEvaluationDatasetsListTable = ({
+  // Set the selected dataset if we don't already have one,
   // or if the selected dataset went out of scope (e.g. was deleted / not in search)
-  if (!selectedDataset || !datasets.some((d) => d.dataset_id === selectedDataset.dataset_id)) {
+  if (datasets?.length && (!selectedDatasetId || !datasetIds.includes(selectedDatasetId))) {
```
not technically a problem with this PR, but can we put this whole state-setting block in an useeffect? otherwise this has risk of very easily causing infinite render loops

```
  useEffect(() => {
    // No datasets -> clear selection
    if (!datasets?.length) {
      setSelectedDataset(undefined);
      setSelectedDatasetId(undefined);
      return;
    }

    // Valid selectedDatasetId from URL -> use it
    if (selectedDatasetId) {
      const matchingDataset = datasets.find((d) => d.dataset_id === selectedDatasetId);
      if (matchingDataset) {
        setSelectedDataset(matchingDataset);
        return;
      }
    }

    // No valid selection -> auto-select first
    const sortedRows = table.getRowModel().rows;
    if (sortedRows.length > 0) {
      const firstDataset = sortedRows[0].original;
      setSelectedDataset(firstDataset);
      setSelectedDatasetId(firstDataset.dataset_id);
    }
  }, [datasets, selectedDatasetId, setSelectedDataset, setSelectedDatasetId, table]);

```

dbczumar replied:
Totally - done.

> otherwise this has risk of very easily causing infinite render loops

I actually observed this while iterating on the PR - glad there's a better way :)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

daniellok-db commented on mlflow/server/js/src/experiment-tracking/pages/experiment-evaluation-datasets/ExperimentEvaluationDatasetsPage.tsx:
```diff
@@ -19,6 +20,7 @@ const ExperimentEvaluationDatasetsPageImpl = () => {
   const [datasetListHidden, setDatasetListHidden] = useState(false);
   const [selectedDataset, setSelectedDataset] = useState<EvaluationDataset | undefined>(undefined);
   const [isDatasetsLoading, setIsDatasetsLoading] = useState(false);
+  const [selectedDatasetId, setSelectedDatasetId] = useSelectedDatasetBySearchParam();
```
rather than using the hook here and passing it down to `ExperimentEvaluationDatasetsListTable`, can we just use it in `ExperimentEvaluationDatasetsListTable` directly?

daniellok-db replied:
oh i think i see what the intention was. can we define a callback that combines `setSelectedDataset` and `setSelectedDatasetId` and pass that in instead (similar to what we did in https://github.com/mlflow/mlflow/pull/19142/files#diff-e2232542c52cc97ec1dd208f20ebfc9ded43289282fa3c8f378b8aa1a318445bR109:

```suggestion
  const [selectedDatasetId, setSelectedDatasetId] = useSelectedDatasetBySearchParam();
  
  const onSelectDataset = useCallback((dataset) => {
    setSelectedDataset(dataset);
    setSelectedDatasetId(dataset.dataset_id);
  });
```

daniellok-db commented on mlflow/server/js/src/experiment-tracking/pages/experiment-evaluation-datasets/components/ExperimentEvaluationDatasetsListTable.tsx:
```diff
@@ -215,19 +213,36 @@ export const ExperimentEvaluationDatasetsListTable = ({
+      if (datasetFromId) {
+        setSelectedDataset(datasetFromId);
+      }
+    }
```
actually thinking about it further, i think we should just remove `selectedDataset` as a state field entirely. it should just purely be derived from `datasets` + `selectedDatasetId`. it's weird to have two sources of truth here and have to keep them in sync.

can we lift the `useSearchEvaluationDatasets` call (lines 177 - 185 above) to the parent component, and pass `datasets` in here? then we can do something like `const selectedDataset = datasets.find(dataset => dataset.dataset_id === selectedDatasetId)` rather than having it be in state

dbczumar committed:
Address PR review feedback for dataset URL sharing

- Rename query param key to 'selectedDatasetId' for clarity
- Lift useSearchEvaluationDatasets to parent component
- Derive selectedDataset from datasets + selectedDatasetId
- Remove duplicate state, simplify auto-selection logic

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/server/js/src/experiment-tracking/pages/experiment-evaluation-datasets/hooks/useSelectedDatasetBySearchParam.ts:
```diff
@@ -0,0 +1,43 @@
+import { useCallback } from 'react';
+import { useSearchParams } from '@mlflow/mlflow/src/common/utils/RoutingUtils';
+
+const QUERY_PARAM_KEY = 'selectedDataset';
```
Done :)

dbczumar committed:
Move auto-selection logic to parent component

Address review comment to make child component purely presentational
by moving the auto-selection useEffect to the parent component.

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Update test to use renamed query param key 'selectedDatasetId'

Signed-off-by: dbczumar <corey.zumar@databricks.com>

daniellok-db reviewed (approved):
Lg!
